### PR TITLE
platform checks: Add a scenario that restarts Redpanda

### PR DESCRIPTION
**Pull request by philip-stoev**
### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by philip-stoev** [comment]
Blocked by https://github.com/MaterializeInc/materialize/issues/14098

**Comment by benesch** [comment]
I pushed up a fix for #14098 directly to this PR. That way we won't be able to call #14098 closed without getting this test fixed.

@cjubb39 @bkirwi—could you please review? Details in commit messages, which I've reproduced below:

----

**storage: remove exactly_once bool from Kafka sinks that is always true**

In Materialize Platform, sinks *always* have EOS. Remove some old code
that supported non-EOS mode.

This makes the logic in the Kafka sink implementation a smidge easier to
follow.

----

**storage: make kafka sink more aggressive about crashing on errors**

Since PR https://github.com/MaterializeInc/materialize/pull/14965, our approach to error handling in the Kafka sink is to
crash on unrecoverable errors, letting Kubernetes restart the sink from
scratch. This is not the most efficient approach, as restarting a sink
from scratch is a bit slow, but it is correct and simple to code.

Unfortunately there were still many vestiges of the old approach to
error handling, where, on an unrecoverable error, the sink would log
message to stderr and then drop itself.

This commit eliminates all those vestigial paths. Errors are either
retried or turned into panics as close as possible to where they are
produced.

To enforce that future code follows the new crash-oriented approach to
error handling, the shutdown_flag bool, which causes the sink to drop
itself, is removed from KafkaSinkState, so that flag can only be raised
by the KafkaSinkToken when the user has dropped the sink, rather than by
any method on KafkaSinkState that encounters an unrecoverable error.

Note that the cleanups here are intentionally prioritizing short term
clarity over the long term goal state of this code. In the long term, we
*do* want some method to bubble up unrecoverable errors to the top level
of the sink implementation, where the sink can be rebuilt without
rebooting the whole process. However, for the moment, I found the
multiple layers of retries and stacked shutdown bools in
retry_on_txn_error too hard to reason about. Ideally, when we add the
graceful retry logic , we can find some way to structure the code that's
easier to follow; perhaps the new timely async bridge will help.

Note that this commit also uses a backoff clamp of 30s throughout all
retry loops in the Kafka sink. The old code used 60s in one place and
10m in another; both feel a bit too long to me.

Fix https://github.com/MaterializeInc/materialize/issues/14098.

**Comment by philip-stoev** [comment]
All tests pass , so please review the rest of the PR and approve so that I can merge. Thank you!

**Review by bkirwi — APPROVED** [review]
I think the test looks good, the code changes are overall clarifying, and the proof is certainly in the pudding.

Maybe I'm a little slow this morning, but I'm having trouble spotting which change in that commit is fixing the restart issue -- most cases that now panic ought to have panicked in the past iiuc, just a little further up the stack. What am I missing?

**Comment by benesch** [comment]
> What am I missing?

I didn't actually prove this in my repro, but I saw at least one code path that could cause `shutdown_flag` to get set without causing a panic. `maybe_emit_progress` is called here...

https://github.com/MaterializeInc/materialize/blob/f06db04739f78ce3eba0db77f8ec493be8103d5a/src/storage/src/sink/kafka.rs#L1323-L1337

without panicking if it returns an error. But `maybe_emit_progress` calls `retry_on_txn_error`, which will have set `shutdown_flag` if there was an error. And so you can end up in a situation where you shut down the sink without panicking. Communicating the error both via the return value and the shutdown bool seemed like the kind of footgun that would bite us repeatedly.

File: src/storage/src/sink/kafka.rs
Comment (cjubb39) [inline src/storage/src/sink/kafka.rs:548]: I believe we want this to be `Ok(_)`, break the loop and can return an error -- or panic, depending how aggressive we want to be.  (@bkirwi thoughts?).

I don't think it's generally safe to just retry whatever operation after we abort the transaction!

File: src/storage/src/sink/kafka.rs
Comment (cjubb39) [inline src/storage/src/sink/kafka.rs:612]: sure -- this is simpler!

File: src/storage/src/sink/kafka.rs
Comment (cjubb39) [inline src/storage/src/sink/kafka.rs:818]: similar suggestion here! `let () = self.send(record).await;`? so we fail to compile if `send` changes to return a `Result` / some information on the `Ok` case.  I have the same suggestion for the other `send` calls throughout.

File: src/storage/src/sink/kafka.rs
Comment (cjubb39) [inline src/storage/src/sink/kafka.rs:899]: same for this function (which is perhaps complicated if we choose to return an error from `retry_on_txn_error` instead of panicking).  I think this is a spot where the `let () = ...` makes it extra clear to the reader that we're not accidentally throwing away an error.

The function has quite a few calls that I think would benefit in the same way

**Review by cjubb39 — COMMENTED** [review]
To summarize my comments:
- I don't think we want to continue to retry in `retry_on_txn_error` after aborting a transaction.  We will likely be in an invalid state
- Especially because we expect to put some more errors back in, I think we ought to litter a bunch of `let () = ...` throughout the code so that the compiler yells at us louder if/when we change the return type of those functions.  This has the added benefit of making it more clear to the reader that we're not just throwing away errors

I think this simplifies the error paths, making them more understandable so we can surface errors more easily and thoroughly in the future -- so in general I'm v happy with the change!

File: src/storage/src/sink/kafka.rs
Comment (bkirwi) [reply src/storage/src/sink/kafka.rs:548]: Yeah, agree this is a bit nervewracking. This happens to give fine behaviour at the moment IIUC (`retry_on_txn_error` is only called for `init_transactions`, `commit_transaction`, and `begin_transaction`; I think the first two will return non-retriable errors if the transaction is aborted, and `begin_transaction` will do the right thing) but it's not a safe assumption in general.

**Comment by benesch** [comment]
Would one of you be up for taking over the PR? I’ve got my hands full
today!

**Comment by bkirwi** [comment]
> I didn't actually prove this in my repro, but I saw at least one code path that could cause shutdown_flag to get set without causing a panic.

Ah, yes, of course -- thanks!

**Comment by cjubb39** [comment]
Two changes for @bkirwi review:
- I panicked if we hit an error that requires aborting the tx -- so that we just start from the top again.  I didn't rip out the whole retrying the abort code since we'll want that back eventually for sure
- I noticed that `flush` can't actually fail so I encoded that in the types -- and as a result pulled back on the `let () = ...` because there are basically no errors that get passed around at this point

**Review by bkirwi — APPROVED** [review]
If the tests are happy, I am happy!

**Comment by cjubb39** [comment]
@philip-stoev this is ready for merging! I can click the button myself tomorrow morning (ET) -- didn't want to do it now in case you had some other tests you wanted to run before merging

**Comment by benesch** [comment]
I'm going to merge since I don't want this to miss the release tomorrow! @philip-stoev do let us know if you observe any more test failures!

**Pull request merged**
